allennlp/training/metrics/categorical_accuracy.py

```diff
@@ -44,18 +52,29 @@ def __call__(self,
             raise ConfigurationError("A gold label passed to Categorical Accuracy contains an id >= {}, "
                                      "the number of classes.".format(num_classes))
 
-        # Top K indexes of the predictions (or fewer, if there aren't K of them).
-        # Special case topk == 1, because it's common and .max() is much faster than .topk().
-        if self._top_k == 1:
-            top_k = predictions.max(-1)[1].unsqueeze(-1)
-        else:
-            top_k = predictions.topk(min(self._top_k, predictions.shape[-1]), -1)[1]
+        predictions = predictions.view((-1, num_classes))
+        gold_labels = gold_labels.view(-1).long()
+        if not self._tie_break:
+            # Top K indexes of the predictions (or fewer, if there aren't K of them).
+            # Special case topk == 1, because it's common and .max() is much faster than .topk().
+            if self._top_k == 1:
+                top_k = predictions.max(-1)[1].unsqueeze(-1)
+            else:
+                top_k = predictions.topk(min(self._top_k, predictions.shape[-1]), -1)[1]
 
-        # This is of shape (batch_size, ..., top_k).
-        correct = top_k.eq(gold_labels.long().unsqueeze(-1)).float()
+            # This is of shape (batch_size, ..., top_k).
+            correct = top_k.eq(gold_labels.unsqueeze(-1)).float()
+        else:
+            # prediction is correct if gold label falls on any of the max scores. distribute score by tie_counts
+            max_predictions = predictions.max(-1)[0]
+            max_predictions_mask = predictions.eq(max_predictions.unsqueeze(-1).expand_as(predictions))
+            correct = max_predictions_mask[torch.arange(gold_labels.numel()).long(), gold_labels].float()
```

Comment:
This line would be clearer as:

```
correct = torch.index_select(max_predictions_mask, 1, gold_labels).float()
```

Reply:
Actually, it is not the same thing. This operation checks if gold_label for an instance lies among the max prediction classes. Lets say `max_predictions` is `batch_size X num_classes` shape and `gold_labels` is `batch_size` shape. `max_predictions` would have 1s in last dimension for classes which have same as max prediction and 0s otherwise. I need to check if indices in gold label are among the max predicted classes or not.

`ith` entry in `gold_labels` has indices on vector of `ith` row in `max_predictions`. I need to index on 1st dimension and 2nd dimension together and so extra `torch.arrange(..)` in first dimension. What you suggest will return me `batch_size X batch_size` tensor but I need `batch_size` tensor.  

Still if I need to use, I need to add extra `diag()`.
```
# Following does work and passes tests though
correct = torch.index_select(max_predictions_mask, 1, gold_labels).diag().float()
```
This looks cleaner, but I has some redundance. What do you suggest?

Reply:
Ok your solution is better, but can you add a comment because it's a pretty complicated bit of indexing.

Reply:
Sure, I will add that.